lib/charliecloud.py

```diff
@@ -979,21 +987,26 @@ def defaults_add(self):
       if (self.tag is None and self.digest is None): self.tag = "latest"
 
    def from_tree(self, t):
-      self.host = tree_child_terminal(t, "ir_hostport", "IR_HOST")
-      self.port = tree_child_terminal(t, "ir_hostport", "IR_PORT")
+      self.host = self.var_sub(tree_child_terminal(t, "ir_hostport", "IR_HOST"))
+      self.port = self.var_sub(tree_child_terminal(t, "ir_hostport", "IR_PORT"))
       if (self.port is not None):
          self.port = int(self.port)
-      self.path = list(tree_child_terminals(t, "ir_path", "IR_PATH_COMPONENT"))
-      self.name = tree_child_terminal(t, "ir_name", "IR_PATH_COMPONENT")
-      self.tag = tree_child_terminal(t, "ir_tag", "IR_TAG")
-      self.digest = tree_child_terminal(t, "ir_digest", "HEX_STRING")
+      for s in tree_child_terminals(t, "ir_path", "IR_PATH_COMPONENT"):
+         self.path.append(self.var_sub(s))
+      self.name = self.var_sub(tree_child_terminal(t, "ir_name", "IR_PATH_COMPONENT"))
+      self.tag = self.var_sub(tree_child_terminal(t, "ir_tag", "IR_TAG"))
+      self.digest = self.var_sub(tree_child_terminal(t, "ir_digest", "HEX_STRING"))
       # Resolve grammar ambiguity for hostnames w/o dot or port.
       if (    self.host is not None
           and "." not in self.host
           and self.port is None):
          self.path.insert(0, self.host)
          self.host = None
 
+   def var_sub(self, var):
+      if (var is not None and var[0] == "$"):
+         return variables_sub(var, self.variables)
+      return var
```

Comment:
If this method isn't called anywhere else, it could be a function defined within `from_tree()`.